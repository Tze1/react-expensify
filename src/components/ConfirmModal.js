import React from 'react';
import { Modal, Button } from 'react-bootstrap';

/* Usage:
* import ConfirmModel from '<path-from-importer-to-ConfirmModal>';
* <Modal
*   show={this.state.showConfirmModal}
*   title='<your-optional-title>'
*   content='<your-required-content>'
*   btnLabelCancel='<your-optional-cancel-btn-label>'
*   btnLabelOk='<your-optional-ok-btn-label>'
*   handleCancel={this.handleCancel}
*   handleOk={this.handleOk}
* />
*/

const ConfirmModal = (props) => (
    <Modal show={ props.show } className="confirmmodal component">
      <Modal.Header className="modal-header">
        <Modal.Title className="modal-title">{ props.title ? props.title : 'Confirm'}</Modal.Title>
      </Modal.Header>

      <Modal.Body className="modal-body">
        <p className="modal-message">{ props.content }</p>
      </Modal.Body>

      <Modal.Footer className="modal-footer">
        <Button
          onClick={ props.handleCancel }
          className="modal-cancel button exit"
        >
          { props.btnLabelCancel ? props.btnLabelCancel : 'Cancel' }
        </Button>
        <Button
          onClick={ props.handleOk }
          className="modal-ok button danger"
        >
          { props.btnLabelOk ? props.btnLabelOk : 'OK' }
        </Button>
      </Modal.Footer>
    </Modal>
);

export default ConfirmModal;
