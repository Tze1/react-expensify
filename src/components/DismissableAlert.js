import React from 'react';
import { connect } from 'react-redux';
import { Alert } from 'react-bootstrap';
import { dismissMessage } from '../actions/messageActions';

export class DismissableAlert extends React.Component {
  constructor (props, context) {
    super(props, context);

    this.handleDismiss = this.handleDismiss.bind(this);
    this.timeoutDismiss = this.timeoutDismiss.bind(this);
  }

  componentDidUpdate () {
    if (this.props.autoDismiss) {
      this.timeoutDismiss();
    }
  }

  handleDismiss () {
    this.props.dismissMessage();
  }

  timeoutDismiss (seconds = 3.5) {
    const vm = this;

    setTimeout(() => {
      // User might have already clicked close-button in the interim
      if (vm.props.show) {
        vm.props.dismissMessage();
      }
    }, seconds * 1000);
  }

  render () {
    if (this.props.show) {
      return (
        <Alert
        bsStyle={this.props.bsStyle}
        onDismiss={this.handleDismiss}
        className="dismissablealert component"
        >
          <h4>{this.props.content}</h4>
        </Alert>
      );
    }
    
    return null
  }
}

const mapStateToProps = (state) => ({
  show: state.message.show,
  bsStyle: state.message.bsStyle,
  content: state.message.content,
  autoDismiss: state.message.autoDismiss,
});

const mapDispatchToProps = (dispatch) => ({
  dismissMessage: () => { dispatch(dismissMessage()) },
});

export default connect(mapStateToProps, mapDispatchToProps)(DismissableAlert);
