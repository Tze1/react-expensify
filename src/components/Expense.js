import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const Expense = (props) => (
  <div id={`expense-${props.id}`} className="expense component">
    <span>{moment(props.createdAt).format('MM/DD/YYYY')}:</span>&nbsp;
    <span>{props.description}</span>&nbsp;
    <span>{numeral(props.amount / 100).format('$0,0.00')}</span>&nbsp;
    <Link to={`/edit/${props.id}`}>Edit</Link>
  </div>
);

export default Expense;
