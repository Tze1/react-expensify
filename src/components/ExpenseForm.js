import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      calendarFocused: false,
      errorMessage: '',
    };

    this.onDescriptionChange = this.onDescriptionChange.bind(this);
    this.onAmountChange = this.onAmountChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onCalendarFocusChange = this.onCalendarFocusChange.bind(this);
    this.onNoteChange = this.onNoteChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onDescriptionChange (e) {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange (e) {
    const amount = e.target.value,
      amtRegex = /^\d{1,}(\.\d{0,2})?$/;
    if (!amount || amount.match(amtRegex)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange (createdAt) {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }
  onCalendarFocusChange ({ focused }) {
    this.setState(() => ({ calendarFocused: focused }));
  }

  onNoteChange (e) {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  onSubmit (e) {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      // Set error message.
      this.setState(() => ({ errorMessage: 'ERROR: Missing description or amount.' }));
    } else {
      this.setState(() => ({ errorMessage: '' }));
      this.props.onSubmit({
        createdAt: this.state.createdAt.valueOf(),
        description: this.state.description,
        amount: Math.round(parseFloat(this.state.amount, 10) * 100),
        note: this.state.note,
      });
    }
  }

  render () {
    return (
      <div className="expenseform component">
        {this.state.errorMessage && <p className="expenseform-error">{this.state.errorMessage}</p>}
        <form onSubmit={this.onSubmit}>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <input
            className="expenseform-description"
            type="text"
            name="description"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            className="expenseform-amount"
            type="text"
            name="amount"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <textarea
            className="expenseform-note"
            name="note"
            placeholder="Add a note.  (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button className="expenseform-submit button" type="submit">{this.props.expense ? 'Update' : 'Save'} Expense</button>
        </form>
      </div>
    );
  }
};
