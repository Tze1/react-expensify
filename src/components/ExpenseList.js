import React from 'react';
import { connect } from 'react-redux';
import Expense from './Expense';
import { filterExpenses } from '../selectors/expensesSelectors';

const ExpenseList = (props) => (
  <div>
    <h3>Expense List</h3>
    {props.expenses.map((expense, index) => {
      return <Expense key={index} {...expense} />;
    })}
  </div>
);

// Export connected component
const mapStateToProps = (state, props) => {
  return {
    expenses: filterExpenses(state.expenses, state.filters)
  }
};
export default connect(mapStateToProps)(ExpenseList);
