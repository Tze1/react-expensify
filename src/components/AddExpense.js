import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expensesActions';

const AddExpense = (props) => (
  <div>
    <h2>Add Expense</h2>
    <ExpenseForm onSubmit={(expense) => {
        props.startAddExpense(expense);
        props.history.push('/');
      }}
    />
    <button onClick={() => props.history.push('/')}>Cancel</button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpense);
