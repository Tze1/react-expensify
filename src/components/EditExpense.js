import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expensesActions';

export class EditExpense extends React.Component {
  constructor (props) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
    this.onRemove = this.onRemove.bind(this);
  }

  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render () {
    return (
      <div className="editexpense component">
        <h2 className="editexpense-title">Edit Expense</h2>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <div className="editexpense-actions">
          <button className="editexpense-cancel caution" onClick={() => this.props.history.push('/')}>Cancel</button>&nbsp;
          <button className="editexpense-remove danger" onClick={this.onRemove}>Remove Expense</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(item => item.id === props.match.params.id)
});
const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);