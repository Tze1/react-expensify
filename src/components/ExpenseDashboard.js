import React from 'react';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseList from './ExpenseList';

const ExpenseDashboard = () => (
  <div>
    <h2>Dashboard</h2>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
