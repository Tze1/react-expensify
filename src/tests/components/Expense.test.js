import React from 'react';
import { shallow } from 'enzyme';
import { Expense } from '../../components/Expense';
import ConfirmModal from '../../components/ConfirmModal';
import testExpenses from '../fixtures/testExpenses';

const testExpense = testExpenses[0];
const startRemoveExpense = jest.fn();
const wrapper = shallow(<Expense {...testExpenses[0] } startRemoveExpense={startRemoveExpense} />);

test('Should render Expense with provided props', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should show remove-confirmation modal on remove-button click', () => {
  wrapper.setState({ showConfirmModal: false });
  wrapper.find('button.expense-remove').simulate('click');

  expect(wrapper.state('showConfirmModal')).toBeTruthy();

  wrapper.setState({ showConfirmModal: false });
});

test('Should hide remove-confirmation modal when cancelled', () => {
  startRemoveExpense.mockClear();
  wrapper.setState({ showConfirmModal: false });

  wrapper.find('button.expense-remove').simulate('click');
  wrapper.instance().onRemoveCancel();

  expect(wrapper.state('showConfirmModal')).toBeFalsy();
  expect(startRemoveExpense).not.toHaveBeenCalled();
})

test('Should handle confirmed removal properly', () => {
  startRemoveExpense.mockClear();
  wrapper.instance().startRemove();

  expect(wrapper.state('showConfirmModal')).toBeFalsy();
  expect(startRemoveExpense).toHaveBeenCalledWith({ id: testExpense.id });
});
