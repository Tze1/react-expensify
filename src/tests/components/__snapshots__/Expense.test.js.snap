// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Expense with provided props 1`] = `
<div
  className="expense component"
  id="expense-1"
>
  <div
    className="expense-createdat"
  >
    12/31/1969
    :
  </div>
  <div
    className="expense-description"
  >
    Gum
  </div>
  <div
    className="expense-amount"
  >
    $1.95
  </div>
  <Link
    className="expense-edit"
    replace={false}
    to="/edit/1"
  >
    edit
  </Link>
</div>
`;
